File: backend/python/src/data_go_kr/test_bid_api.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import requests
import json
import os
import sys
from datetime import datetime, timedelta
from dotenv import load_dotenv

# utils_mysql 모듈 import를 위한 경로 추가
sys.path.append('/exposed/projects/bid-notice-web/backend/python/src')
from utils.utils_mysql import Mysql

def load_service_key():
    """Load service key from .env file"""
    env_path = os.path.join(os.path.dirname(__file__), '.env')
    load_dotenv(env_path)
    return os.getenv('DATA_GO_KR_SERVICE_KEY')

def save_response_to_file(data):
    """Save API response to JSON file with timestamp"""
    now = datetime.now()
    filename = f"res_{now.strftime('%m%d_%H_%M_%S')}.json"

    # Create responses directory if it doesn't exist
    # responses_dir = os.path.join(os.path.dirname(__file__), 'responses')
    responses_dir = '/exposed/projects/bid-notice-web/backend/python/src/data_go_kr/responses'
    os.makedirs(responses_dir, exist_ok=True)

    # Full file path
    file_path = os.path.join(responses_dir, filename)

    try:
        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        print(f"Response saved to: {file_path}")
    except Exception as e:
        print(f"Error saving response to file: {e}")

def camel_to_snake(name):
    """Convert camelCase to snake_case"""
    import re
    s1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()

def convert_keys_to_snake_case(data_dict):
    """Convert all dictionary keys from camelCase to snake_case"""
    return {camel_to_snake(k): v for k, v in data_dict.items()}

def safe_val_str(val):
    """값을 MySQL 쿼리에서 사용할 수 있는 형식으로 안전하게 변환"""
    if val is None:
        return "NULL"
    elif isinstance(val, str):
        return "'" + val.replace("'", "\\'").replace('"', '\\"') + "'"
    else:
        return str(val)

def save_to_mysql(data):
    """Save API response data to MySQL notices_g2b table"""
    try:
        mysql = Mysql()
        
        # response > body > items 배열에서 데이터 추출
        if 'response' not in data or 'body' not in data['response'] or 'items' not in data['response']['body']:
            print("No items found in API response")
            return
        
        items = data['response']['body']['items']
        if not items:
            print("Empty items array")
            return
        
        print(f"Processing {len(items)} items...")
        
        for item in items:
            # 카멜케이스를 스네이크케이스로 변환
            snake_case_item = convert_keys_to_snake_case(item)
            
            # None 키 제거 및 빈 문자열 처리
            cleaned_item = {}
            for key, value in snake_case_item.items():
                # None 키나 빈 키 제거
                if key is None or key == '':
                    continue
                    
                # 빈 문자열을 None으로 변환
                if value == "":
                    cleaned_item[key] = None
                else:
                    cleaned_item[key] = value
            
            # 수동으로 INSERT 쿼리 생성 (None 값 처리를 위해)
            try:
                keys = list(cleaned_item.keys())
                values = list(cleaned_item.values())
                
                keys_str = ", ".join([f"`{k}`" for k in keys])
                vals_str = ", ".join([safe_val_str(v) for v in values])
                
                sql = f"INSERT IGNORE INTO notices_g2b ({keys_str}) VALUES ({vals_str});"
                mysql.exec(sql)
                
                print(f"Inserted: {cleaned_item.get('bid_ntce_no', 'N/A')} - {cleaned_item.get('bid_ntce_nm', 'N/A')[:50]}...")
            except Exception as e:
                print(f"Error inserting item {cleaned_item.get('bid_ntce_no', 'N/A')}: {e}")
                continue
        
        print("MySQL save completed")
        
    except Exception as e:
        print(f"Error saving to MySQL: {e}")
    finally:
        if 'mysql' in locals():
            mysql.close()

def test_bid_pblnc_list_info_service():
    """Test getBidPblancListInfoServc API"""
    
    # Load service key
    service_key = load_service_key()
    if not service_key:
        print("Error: SERVICE_KEY not found in .env file")
        return None
    
    # API endpoint
    base_url = "https://apis.data.go.kr/1230000/ad/BidPublicInfoService"
    endpoint = f"{base_url}/getBidPblancListInfoServc"
    
    # Calculate date range (last 3 days)
    end_date = datetime.now()
    start_date = end_date - timedelta(days=3)
    
    # Parameters - try different date format
    params = {
        'serviceKey': service_key,
        'pageNo': '1',
        'numOfRows': '10',  # Get 10 records for testing
        'type': 'json',     # Response format
        'inqryDiv': '1',    # 조회구분 (1: 공고일자)
        'inqryBgnDt': start_date.strftime('%Y%m%d%H%M'),  # YYYYMMDDHHMM 형식
        'inqryEndDt': end_date.strftime('%Y%m%d%H%M')     # YYYYMMDDHHMM 형식
    }
    
    print("=== 공공데이터포털 입찰공고 API 테스트 ===")
    print(f"Endpoint: {endpoint}")
    print(f"Parameters: {json.dumps(params, indent=2, ensure_ascii=False)}")
    print()
    
    try:
        # Make API request
        print("Making API request...")
        response = requests.get(endpoint, params=params, timeout=30)
        
        print(f"Status Code: {response.status_code}")
        print(f"Response Headers: {dict(response.headers)}")
        print()
        
        if response.status_code == 200:
            try:
                # Parse JSON response
                data = response.json()
                
                print("=== API Response ===")
                print(json.dumps(data, indent=2, ensure_ascii=False))
                
                # Extract and display key information
                if 'header' in data:
                    header = data['header']
                    print(f"\n=== Header Info ===")
                    print(f"Result Code: {header.get('resultCode', 'N/A')}")
                    print(f"Result Message: {header.get('resultMsg', 'N/A')}")
                
                if 'body' in data and data['body']:
                    body = data['body']
                    print(f"\n=== Body Info ===")
                    print(f"Page Number: {body.get('pageNo', 'N/A')}")
                    print(f"Number of Rows: {body.get('numOfRows', 'N/A')}")
                    print(f"Total Count: {body.get('totalCount', 'N/A')}")
                    
                    # Display items if available
                    if 'items' in body and body['items']:
                        items = body['items']
                        print(f"\n=== Items Info ===")
                        
                        if isinstance(items.get('item'), list):
                            print(f"Number of items: {len(items['item'])}")
                            for i, item in enumerate(items['item'], 1):
                                print(f"\n--- Item {i} ---")
                                print(f"입찰공고번호: {item.get('bidNtceNo', 'N/A')}")
                                print(f"입찰공고명: {item.get('bidNtceNm', 'N/A')}")
                                print(f"공고기관명: {item.get('ntceInsttNm', 'N/A')}")
                                print(f"수요기관명: {item.get('dminsttNm', 'N/A')}")
                                print(f"입찰공고일: {item.get('bidNtceDt', 'N/A')}")
                                print(f"입찰마감일시: {item.get('bidClseDt', 'N/A')}")
                                print(f"개찰일시: {item.get('opengDt', 'N/A')}")
                                print(f"예정가격: {item.get('presmptPrce', 'N/A')}")
                                print(f"공고구분명: {item.get('ntceKindNm', 'N/A')}")
                        elif isinstance(items.get('item'), dict):
                            item = items['item']
                            print(f"Single item found:")
                            print(f"입찰공고번호: {item.get('bidNtceNo', 'N/A')}")
                            print(f"입찰공고명: {item.get('bidNtceNm', 'N/A')}")
                            print(f"공고기관명: {item.get('ntceInsttNm', 'N/A')}")
                            print(f"수요기관명: {item.get('dminsttNm', 'N/A')}")
                        else:
                            print("No items found in response")
                    else:
                        print("No items found in response")
                
                # Save response to file
                save_response_to_file(data)
                
                # Save to MySQL database
                save_to_mysql(data)
                
                return data
                
            except json.JSONDecodeError as e:
                print(f"Error parsing JSON response: {e}")
                print(f"Raw response: {response.text[:1000]}...")
                return None
        else:
            print(f"Error: HTTP {response.status_code}")
            print(f"Response: {response.text}")
            return None
            
    except requests.exceptions.RequestException as e:
        print(f"Request error: {e}")
        return None

if __name__ == "__main__":
    test_bid_pblnc_list_info_service()

File: backend/nodejs/src/utils/nara/tests/test-bid-api.ts
/**
 * Test script for Korean Government Bid Notice API
 * Equivalent to the Python test_bid_api.py
 */

import { config } from 'dotenv';
import { writeFile, mkdir } from 'fs/promises';
import { existsSync } from 'fs';
import path from 'path';
import mysql from 'mysql2/promise';
import { DataGoKrApiClient, BidNoticeRawItem } from './api-client.js';
import { BidNoticeParser } from './parser.js';
import { BidNotice } from './models.js';

// Load environment variables
config();

/**
 * Load service key from environment
 */
function loadServiceKey(): string {
  const serviceKey = process.env.DATA_GO_KR_SERVICE_KEY;
  if (!serviceKey) {
    console.error('Error: DATA_GO_KR_SERVICE_KEY not found in .env file');
    process.exit(1);
  }
  return serviceKey;
}

/**
 * Save API response to JSON file with timestamp
 */
async function saveResponseToFile(data: any): Promise<void> {
  const now = new Date();
  const timestamp = now.toISOString()
    .replace(/[:.]/g, '_')
    .replace('T', '_')
    .slice(0, -5); // Remove milliseconds and Z

  const filename = `res_${timestamp}.json`;

  // Create responses directory if it doesn't exist
  const responsesDir = path.join(process.cwd(), 'responses');
  if (!existsSync(responsesDir)) {
    await mkdir(responsesDir, { recursive: true });
  }

  // Full file path
  const filePath = path.join(responsesDir, filename);

  try {
    await writeFile(filePath, JSON.stringify(data, null, 2), 'utf8');
    console.log(`Response saved to: ${filePath}`);
  } catch (error: any) {
    console.error(`Error saving response to file: ${error.message}`);
  }
}

/**
 * Convert camelCase to snake_case
 */
function camelToSnake(name: string): string {
  return name
    .replace(/([a-z\d])([A-Z])/g, '$1_$2')
    .replace(/([A-Z]+)([A-Z][a-z\d]+)/g, '$1_$2')
    .toLowerCase();
}

/**
 * Convert all dictionary keys from camelCase to snake_case
 */
function convertKeysToSnakeCase(obj: Record<string, any>): Record<string, any> {
  const result: Record<string, any> = {};
  for (const [key, value] of Object.entries(obj)) {
    result[camelToSnake(key)] = value;
  }
  return result;
}

/**
 * Safe value string for MySQL queries
 */
function safeValStr(val: any): string {
  if (val === null || val === undefined) {
    return 'NULL';
  } else if (typeof val === 'string') {
    return "'" + val.replace(/'/g, "\\'").replace(/"/g, '\\"') + "'";
  } else {
    return String(val);
  }
}

/**
 * Save API response data to MySQL notices_g2b table
 */
async function saveToMysql(data: any): Promise<void> {
  let connection: mysql.Connection | null = null;

  try {
    // Create MySQL connection
    connection = await mysql.createConnection({
      host: process.env.MYSQL_HOST || 'localhost',
      port: parseInt(process.env.MYSQL_PORT || '3306'),
      user: process.env.MYSQL_USER || 'root',
      password: process.env.MYSQL_PASSWORD || '',
      database: process.env.MYSQL_DATABASE || 'test',
      charset: 'utf8mb4'
    });

    // Extract items from response
    let items: BidNoticeRawItem[] = [];

    if (data?.response?.body?.items) {
      if (Array.isArray(data.response.body.items)) {
        items = data.response.body.items;
      } else if (data.response.body.items.item) {
        items = Array.isArray(data.response.body.items.item)
          ? data.response.body.items.item
          : [data.response.body.items.item];
      }
    } else if (data?.body?.items) {
      items = Array.isArray(data.body.items) ? data.body.items : [data.body.items];
    } else if (Array.isArray(data)) {
      items = data;
    }

    if (!items || items.length === 0) {
      console.log('No items found in API response');
      return;
    }

    console.log(`Processing ${items.length} items...`);

    for (const item of items) {
      try {
        // Clean item (remove null/empty keys) - keep camelCase for g2b_notices table
        const cleanedItem: Record<string, any> = {};
        for (const [key, value] of Object.entries(item)) {
          // Remove null or empty keys
          if (key === null || key === undefined || key === '') {
            continue;
          }

          // Convert empty strings to null
          cleanedItem[key] = value === '' ? null : value;
        }

        // Add default values for new fields
        cleanedItem.category = cleanedItem.category || '무관';
        cleanedItem.is_selected = cleanedItem.is_selected || 0;

        // Generate INSERT query manually (for null value handling)
        const keys = Object.keys(cleanedItem);
        const values = Object.values(cleanedItem);

        const keysStr = keys.map(k => `\`${k}\``).join(', ');
        const valsStr = values.map(v => safeValStr(v)).join(', ');

        const sql = `INSERT IGNORE INTO g2b_notices (${keysStr}) VALUES (${valsStr})`;

        await connection.execute(sql);

        const bidNtceNo = cleanedItem.bidNtceNo || 'N/A';
        const bidNtceNm = cleanedItem.bidNtceNm || 'N/A';
        console.log(`Inserted: ${bidNtceNo} - ${bidNtceNm.substring(0, 50)}...`);

      } catch (error: any) {
        const bidNtceNo = item.bidNtceNo || 'N/A';
        console.error(`Error inserting item ${bidNtceNo}: ${error.message}`);
        continue;
      }
    }

    console.log('MySQL save completed');

  } catch (error: any) {
    console.error(`Error saving to MySQL: ${error.message}`);
  } finally {
    if (connection) {
      await connection.end();
    }
  }
}

/**
 * Test getBidPblancListInfoServc API
 */
async function testBidPblancListInfoService(): Promise<any> {
  // Load service key
  const serviceKey = loadServiceKey();

  // Create API client
  const apiClient = new DataGoKrApiClient(serviceKey);

  // Calculate date range (last 3 days)
  const endDate = new Date();
  const startDate = new Date();
  startDate.setDate(startDate.getDate() - 3);

  // API parameters
  const params = {
    pageNo: 1,
    numOfRows: 10, // Get 10 records for testing
    type: 'json' as const,
    inqryDiv: '1', // 조회구분 (1: 공고일자)
    inqryBgnDt: startDate.toISOString().slice(0, 10).replace(/-/g, '') + '0000', // YYYYMMDD0000
    inqryEndDt: endDate.toISOString().slice(0, 10).replace(/-/g, '') + '0000'     // YYYYMMDD0000
  };

  console.log('=== 공공데이터포털 입찰공고 API 테스트 ===');
  console.log('Parameters:', JSON.stringify(params, null, 2));
  console.log();

  try {
    console.log('Making API request...');

    // Make API request
    const data = await apiClient.getBidPblancListInfoServc(params);

    console.log('=== API Response ===');
    console.log(`Retrieved ${data.length} items`);

    if (data.length > 0) {
      console.log('\n=== Sample Items ===');

      data.slice(0, 3).forEach((item, index) => {
        console.log(`\n--- Item ${index + 1} ---`);
        console.log(`입찰공고번호: ${item.bidNtceNo || 'N/A'}`);
        console.log(`입찰공고명: ${item.bidNtceNm || 'N/A'}`);
        console.log(`공고기관명: ${item.ntceInsttNm || 'N/A'}`);
        console.log(`수요기관명: ${item.dminsttNm || 'N/A'}`);
        console.log(`입찰공고일: ${item.bidNtceDt || 'N/A'}`);
        console.log(`입찰마감일시: ${item.bidClseDt || 'N/A'}`);
        console.log(`개찰일시: ${item.opengDt || 'N/A'}`);
        console.log(`예정가격: ${item.presmptPrce || 'N/A'}`);
      });
    }

    // Test parsing
    console.log('\n=== Testing Parser ===');
    if (data.length > 0) {
      const firstItem = data[0];
      const parsed = BidNoticeParser.parseBidNotice(firstItem);

      console.log('Parsed notice:');
      console.log(`ID: ${parsed.bid_notice_no}`);
      console.log(`Name: ${parsed.bid_notice_name}`);
      console.log(`Dept: ${parsed.dept_name}`);
      console.log(`Notice Date: ${parsed.notice_date?.toISOString()?.split('T')[0]}`);
      console.log(`Budget: ${parsed.budget_amount}`);
    }

    // Save response to file
    await saveResponseToFile(data);

    // Save to MySQL database
    console.log('\n=== Saving to MySQL ===');
    await saveToMysql({ response: { body: { items: data } } });

    return data;

  } catch (error: any) {
    console.error(`Request error: ${error.message}`);
    return null;
  }
}

/**
 * Test complete service integration
 */
async function testServiceIntegration(): Promise<void> {
  console.log('\n=== Testing Service Integration ===');

  const { BidNoticeService } = await import('./service.js');

  const config = {
    serviceKey: loadServiceKey(),
    host: process.env.MYSQL_HOST || 'localhost',
    port: parseInt(process.env.MYSQL_PORT || '3306'),
    user: process.env.MYSQL_USER || 'root',
    password: process.env.MYSQL_PASSWORD || '',
    database: process.env.MYSQL_DATABASE || 'test'
  };

  const service = new BidNoticeService(config);

  try {
    // Test today's collection
    console.log('Testing today\'s collection...');
    const result = await service.collectTodayNotices({
      saveToDatabase: false, // Don't save to avoid table dependency
      applyKeywordMatching: false
    });

    console.log('Service result:');
    console.log(`Success: ${result.success}`);
    console.log(`Total: ${result.collection_result.total_count}`);
    console.log(`Collected: ${result.collection_result.collected_count}`);
    console.log(`Errors: ${result.collection_result.error_count}`);

  } catch (error: any) {
    console.error('Service test failed:', error.message);
  } finally {
    await service.close();
  }
}

/**
 * Main test function
 */
async function main(): Promise<void> {
  const testType = process.argv[2];

  switch (testType) {
    case 'api':
      await testBidPblancListInfoService();
      break;
    case 'service':
      await testServiceIntegration();
      break;
    default:
      console.log('Testing API client...');
      await testBidPblancListInfoService();

      console.log('\nTesting service integration...');
      await testServiceIntegration();
  }
}

// Run if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((error) => {
    console.error('Test failed:', error);
    process.exit(1);
  });
}

export { testBidPblancListInfoService, testServiceIntegration };